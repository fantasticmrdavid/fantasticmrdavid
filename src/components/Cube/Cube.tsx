import React, { memo } from "react";
import { BorderTracer } from "components/BorderTracer/BorderTracer";
import {
  Container,
  Polygon,
  Front,
  Back,
  Left,
  Right,
  Top,
  Bottom,
} from "./cube.styles";

export interface CubeProps {
  isLoading: boolean;
}

export const Cube = memo(({ isLoading }: CubeProps) => (
  <Container $isLoading={isLoading}>
    <Polygon>
      <Front>
        <BorderTracer direction="left" cycleColor />
      </Front>
      <Back>
        <BorderTracer cycleColor />
      </Back>
      <Top />
      <Bottom>
        <BorderTracer cycleColor />
      </Bottom>
      <Left>
        <BorderTracer direction="left" cycleColor />
      </Left>
      <Right>
        <BorderTracer cycleColor />
      </Right>
    </Polygon>
  </Container>
));
