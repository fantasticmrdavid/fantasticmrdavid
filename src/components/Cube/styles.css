.Cube_container {
  position: absolute;
  bottom: 0px;
  right: 0px;
	perspective: 800px;
	perspective-origin: 50% 150px;
  transition: 0.5s;
  transform: scale(2);

  &.loading {
    transform: scale(8);
  }
}

.Cube {
	position: absolute;;
  bottom: 225px;
  right: -100px;
	width: 300px;
	transform-style: preserve-3d;
  animation: spin 0.6s normal linear;
  animation-fill-mode: forwards;

  div {
  	position: absolute;
  	width: 300px;
  	height: 300px;
    background: rgba(209, 43, 43, 0.1);
    border: 1px solid rgba(209, 43, 43, 0.2);
  }

  .back {
  	transform: translateZ(-150px) rotateY(180deg);
  }
  .right {
  	transform: rotateY(-270deg) translateX(150px);
  	transform-origin: top right;
  }
  .left {
  	transform: rotateY(270deg) translateX(-150px);
  	transform-origin: center left;
  }
  .top {
  	transform: rotateX(-90deg) translateY(-150px);
  	transform-origin: top center;
  }
  .bottom {
  	transform: rotateX(90deg) translateY(150px);
  	transform-origin: bottom center;
  }
  .front {
  	transform: translateZ(150px);
  }
}

@keyframes spin {
	from { transform: rotateX(15deg) rotateY(15deg); }
	to { transform: rotateX(-60deg) rotateY(60deg); }
}
