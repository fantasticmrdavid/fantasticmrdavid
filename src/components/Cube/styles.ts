import styled, { css, keyframes } from 'styled-components';
import { transparentize } from 'polished';

const cubeColors = [
  'rgb(209, 43, 43)',
  'rgb(66, 33, 99)',
  'rgb(0, 0, 255)',
];

interface ContainerProps {
  loading: boolean,
}

const spin = keyframes`
  from { transform: rotateX(15deg) rotateY(15deg); }
  to { transform: rotateX(-60deg) rotateY(60deg); }
`;

const cycleColor = keyframes`
  ${cubeColors.map((c: string, i: number) => `
    ${i * (100 / cubeColors.length)}% {
      background-color: ${transparentize(0.9, c)};
      border: ${`1px solid ${transparentize(0.8, c)}`};
    }`)}
    100% {
      background-color: ${transparentize(0.9, cubeColors[0])};
      border: ${`1px solid ${transparentize(0.8, cubeColors[0])}`};
    }
`;

export const Container = styled.div`
  position: absolute;
  bottom: 0px;
  right: 0px;
  perspective: 800px;
  perspective-origin: 50% 150px;
  transition: 0.5s;
  transform: ${css`scale(${(props: ContainerProps) => (props.loading ? 8 : 2)})`};
`;

export const Polygon = styled.div`
  position: absolute;
  bottom: 225px;
  right: -100px;
  width: 300px;
  transform-style: preserve-3d;
  animation: ${css`${spin} 0.6s normal linear`};
  animation-fill-mode: forwards;
`;

const Side = styled.div`
  position: absolute;
  width: 300px;
  height: 300px;
  animation: ${css`${cycleColor} 45s normal infinite`};
`;

export const Back = styled(Side)`
  transform: translateZ(-150px) rotateY(180deg);
`;

export const Right = styled(Side)`
  transform: rotateY(-270deg) translateX(150px);
  transform-origin: top right;
`;

export const Left = styled(Side)`
  transform: rotateY(270deg) translateX(-150px);
  transform-origin: center left;
`;

export const Top = styled(Side)`
  transform: rotateX(-90deg) translateY(-152px);
  transform-origin: top center;
`;

export const Bottom = styled(Side)`
  transform: rotateX(90deg) translateY(150px);
  transform-origin: bottom center;
`;

export const Front = styled(Side)`
  transform: translateZ(150px);
`;
