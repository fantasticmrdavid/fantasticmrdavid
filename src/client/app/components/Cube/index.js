import React, { PropTypes } from 'react';
import styles from './styles.css';

const Cube = (props) => {
  const { loading } = props;
  return (
    <div className={`${styles.Cube_container} ${!!loading && styles.loading}`}>
      <div className={styles.Cube}>
        <div className={styles.front} />
        <div className={styles.back} />
        <div className={styles.top} />
        <div className={styles.bottom} />
        <div className={styles.left} />
        <div className={styles.right} />
      </div>
    </div>
  );
};

Cube.propTypes = {
  loading: PropTypes.bool,
};

export default Cube;
