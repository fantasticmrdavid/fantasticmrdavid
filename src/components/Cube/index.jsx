import React from 'react';
import PropTypes from 'prop-types';
import * as styles from './styles';

const Cube = ({ loading }) => {
  const {
    Container,
    Polygon,
    Front,
    Back,
    Left,
    Right,
    Top,
    Bottom,
  } = styles;

  return (
    <Container loading={loading}>
      <Polygon>
        <Front />
        <Back />
        <Top />
        <Bottom />
        <Left />
        <Right />
      </Polygon>
    </Container>
  );
};

Cube.propTypes = {
  loading: PropTypes.bool,
};

export default Cube;
