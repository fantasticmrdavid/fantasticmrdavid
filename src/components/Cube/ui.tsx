import React, { memo } from 'react';
import BorderTracer from 'components/BorderTracer';
import * as styles from './styles';

interface Props {
  loading: boolean,
}

const {
  Container,
  Polygon,
  Front,
  Back,
  Left,
  Right,
  Top,
  Bottom,
} = styles;

export default memo(
  ({ loading }: Props) => (
    <Container loading={loading}>
      <Polygon>
        <Front><BorderTracer direction="left" cycleColor /></Front>
        <Back><BorderTracer cycleColor /></Back>
        <Top />
        <Bottom><BorderTracer cycleColor /></Bottom>
        <Left><BorderTracer direction="left" cycleColor /></Left>
        <Right><BorderTracer cycleColor /></Right>
      </Polygon>
    </Container>
  ),
);
