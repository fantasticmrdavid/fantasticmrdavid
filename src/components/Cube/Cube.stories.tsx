import React from "react";
import { Cube } from "./Cube";
import { ComponentStory, Meta } from "@storybook/react";

export default {
  title: "Cube",
  component: Cube
} as Meta

const Template: ComponentStory<typeof Cube> = (args) => <div style={{ width: "100%", height: "100%", position: "relative" }}><Cube {...args} /></div>;

export const Loading = Template.bind({ isLoading: true });
export const Loaded = Template.bind({ isLoading: false });
