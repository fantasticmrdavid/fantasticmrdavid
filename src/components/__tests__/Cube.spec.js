import React from 'react';
import renderer from 'react-test-renderer';
import Cube from 'components/Cube';

describe('Component -> Cube', () => {
  it('should render the Cube in loading state without throwing an error', () => {
    expect(renderer.create(<Cube loading={true} />)).toMatchSnapshot();
  });

  it('should render the Cube in completed state without throwing an error', () => {
    expect(renderer.create(<Cube loading={false} />)).toMatchSnapshot();
  });
});
