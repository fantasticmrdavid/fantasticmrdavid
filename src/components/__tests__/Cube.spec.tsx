import React from 'react';
import { render } from "@testing-library/react";
import Cube from 'components/Cube';

describe('Component -> Cube', () => {
  it('should render the Cube in loading state without throwing an error', () => {
    expect(render(<Cube isLoading />)).toMatchSnapshot();
  });

  it('should render the Cube in completed state without throwing an error', () => {
    expect(render(<Cube isLoading={false} />)).toMatchSnapshot();
  });
});
