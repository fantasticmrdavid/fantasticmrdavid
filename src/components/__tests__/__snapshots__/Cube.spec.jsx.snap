// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component -> Cube should render the Cube in completed state without throwing an error 1`] = `
<div
  className="styles__Container-sc-1pwogyu-0 fkagAV"
>
  <div
    className="styles__Polygon-sc-1pwogyu-1 dCIqjj"
  >
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Front-sc-1pwogyu-8 deBIOm"
    >
      <div
        className="styles__Tracer-sc-18vfsm6-0 hNwUQp"
        direction="left"
      />
    </div>
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Back-sc-1pwogyu-3 gCcrQb"
    >
      <div
        className="styles__Tracer-sc-18vfsm6-0 lKbcB"
      />
    </div>
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Top-sc-1pwogyu-6 eLxrDv"
    />
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Bottom-sc-1pwogyu-7 cpagcj"
    >
      <div
        className="styles__Tracer-sc-18vfsm6-0 lKbcB"
      />
    </div>
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Left-sc-1pwogyu-5 iDSTJS"
    >
      <div
        className="styles__Tracer-sc-18vfsm6-0 hNwUQp"
        direction="left"
      />
    </div>
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Right-sc-1pwogyu-4 fTRfub"
    >
      <div
        className="styles__Tracer-sc-18vfsm6-0 lKbcB"
      />
    </div>
  </div>
</div>
`;

exports[`Component -> Cube should render the Cube in loading state without throwing an error 1`] = `
<div
  className="styles__Container-sc-1pwogyu-0 irMsJv"
>
  <div
    className="styles__Polygon-sc-1pwogyu-1 dCIqjj"
  >
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Front-sc-1pwogyu-8 deBIOm"
    >
      <div
        className="styles__Tracer-sc-18vfsm6-0 hNwUQp"
        direction="left"
      />
    </div>
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Back-sc-1pwogyu-3 gCcrQb"
    >
      <div
        className="styles__Tracer-sc-18vfsm6-0 lKbcB"
      />
    </div>
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Top-sc-1pwogyu-6 eLxrDv"
    />
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Bottom-sc-1pwogyu-7 cpagcj"
    >
      <div
        className="styles__Tracer-sc-18vfsm6-0 lKbcB"
      />
    </div>
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Left-sc-1pwogyu-5 iDSTJS"
    >
      <div
        className="styles__Tracer-sc-18vfsm6-0 hNwUQp"
        direction="left"
      />
    </div>
    <div
      className="styles__Side-sc-1pwogyu-2 styles__Right-sc-1pwogyu-4 fTRfub"
    >
      <div
        className="styles__Tracer-sc-18vfsm6-0 lKbcB"
      />
    </div>
  </div>
</div>
`;
